src/agent/onefuzz-agent/src/tasks/heartbeat.rs

```diff
@@ -27,114 +24,52 @@ pub enum HeartbeatData {
-}
-
-impl Drop for HeartbeatClient {
-    fn drop(&mut self) {
-        self.cancelled.notify();
-    }
-}
-
-impl HeartbeatClient {
-    pub fn init(queue_url: Url, task_id: Uuid) -> Self {
-        HeartbeatClient::init_with_period(queue_url, task_id, DEFAULT_HEARTBEAT_PERIOD)
-    }
+pub type TaskHeartbeatClient = HeartbeatClient<HeartbeatData>;
 
-    pub fn init_with_period(queue_url: Url, task_id: Uuid, heartbeat_period: Duration) -> Self {
-        let messages = Arc::new(Mutex::new(HashSet::new()));
-        let cancelled = Arc::new(Notify::new());
-        let _heartbeat_process = HeartbeatClient::start_background_process(
-            task_id,
+pub fn init_task_heartbeat(
+    queue_url: Url,
+    task_id: Option<Uuid>,
+) -> HeartbeatClient<HeartbeatData> {
+    let heartbeat_period = DEFAULT_HEARTBEAT_PERIOD;
+    let messages = Arc::new(Mutex::new(HashSet::<HeartbeatData>::new()));
+    let process_messages = Arc::clone(&messages);
+    let cancelled = Arc::new(Notify::new());
+    let process_cancelled = Arc::clone(&cancelled);
+    let heartbeat_process = task::spawn(async move {
+        let machine_id = get_machine_id().await.unwrap();
+        HeartbeatClient::start_background_process(
             queue_url,
-            messages.clone(),
-            cancelled.clone(),
+            process_messages,
+            process_cancelled,
             heartbeat_period,
-        );
-        HeartbeatClient {
-            messages,
-            _heartbeat_process,
-            cancelled,
-        }
-    }
-
-    fn drain_current_messages(messages: Arc<Mutex<HashSet<HeartbeatData>>>) -> Vec<HeartbeatData> {
-        let lock = messages.lock();
-        let mut messages = lock.unwrap();
-        let drain = messages.iter().cloned().collect::<Vec<HeartbeatData>>();
-        messages.clear();
-        drain
-    }
-
-    async fn flush(
-        task_id: Uuid,
-        machine_id: Uuid,
-        machine_name: &str,
-        queue_client: &QueueClient,
-        messages: Arc<Mutex<HashSet<HeartbeatData>>>,
-    ) {
-        let mut data = HeartbeatClient::drain_current_messages(messages.clone());
-        data.push(HeartbeatData::MachineAlive);
-        let _ = queue_client
-            .enqueue(Heartbeat {
-                task_id,
-                data,
-                machine_id,
-                machine_name,
-            })
-            .await;
-    }
-
-    pub fn start_background_process(
-        task_id: Uuid,
-        queue_url: Url,
-        messages: Arc<Mutex<HashSet<HeartbeatData>>>,
-        cancelled: Arc<Notify>,
-        heartbeat_period: Duration,
-    ) -> JoinHandle<Result<()>> {
-        let queue_client = QueueClient::new(queue_url);
-        task::spawn(async move {
-            let machine_id = get_machine_id().await?;
-            let machine_name = get_machine_name().await?;
+            |queue_client: Arc<QueueClient>, m: Arc<Mutex<HashSet<HeartbeatData>>>| async move {
```

Comment:
Also: this callback body, and the specific type of `HeartbeatData`, are the only things specific to this particular heartbeat client, right? You also have `task_id` in one, but not the other. But, the task ID only gets used in the callback, which closes over it by move.

If it is doable, I'd love to see if if we could make it so users / call sites only have to specify the callback, period.